mlem/contrib/venv.py

```diff
@@ -0,0 +1,173 @@
+import os
+import platform
+import subprocess
+import sys
+import venv
+from abc import abstractmethod
+from typing import ClassVar, Optional
+
+from mlem.core.errors import MlemError
+from mlem.core.objects import MlemBuilder, MlemModel
+from mlem.ui import EMOJI_FAIL, EMOJI_OK, EMOJI_PACK, echo
+
+
+class EnvBuilder(MlemBuilder):
+    """MlemBuilder implementation for building virtual environments"""
+
+    type: ClassVar = "env"
+
+    target: Optional[str] = "venv"
+    """Name of the virtual environment"""
+
+    @abstractmethod
+    def create_virtual_env(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_installed_packages(self, env_dir):
+        raise NotImplementedError
+
+    @abstractmethod
+    def build(self, obj: MlemModel):
+        raise NotImplementedError
+
+
+class VenvBuilder(EnvBuilder):
+
+    type: ClassVar = "venv"
+
+    no_cache: Optional[bool] = False
+    """Disable cache"""
+    current_env: Optional[bool] = False
+    """Whether to install in the current virtual env, must be active"""
+
+    def create_virtual_env(self):
+        env_spec = venv.EnvBuilder(with_pip=True)
+        env_dir = os.path.abspath(self.target)
+
+        context = env_spec.ensure_directories(env_dir)
+
+        true_system_site_packages = env_spec.system_site_packages
+        env_spec.system_site_packages = False
+        env_spec.create_configuration(context)
+        env_spec.setup_python(context)
+        if env_spec.with_pip:
+            env_spec._setup_pip(context)  # pylint: disable=protected-access
+        if not env_spec.upgrade:
+            env_spec.setup_scripts(context)
+            env_spec.post_setup(context)
+        if true_system_site_packages:
+            env_spec.system_site_packages = True
+            env_spec.create_configuration(context)
+
+        return context
+
+    def get_installed_packages(self, env_dir):
+        if platform.system() == "Windows":
+            env_exe = os.path.join(env_dir, "Scripts", "python.exe")
+        else:
+            env_exe = os.path.join(env_dir, "bin", "python")
+
+        return subprocess.check_output([env_exe, "-m", "pip", "freeze"])
```

Comment:
handle errors?

Reply:
For places where we use `subprocess.call`, we don't know what errors can occur so the question is -- which ones do we handle? Or should we just handle a general `Exception` -- but that will be too broad I guess.

Reply:
done, handled 3 kinds -- `FileNotFoundError`, `subprocess.CalledProcessError` and `subprocess.TimeoutExpired`